File: backend/src/main/java/com/pgis/hrms/modules/leave/util/SriLankanHolidays.java
package com.pgis.hrms.modules.leave.util;

import java.time.LocalDate;
import java.time.Month;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Utility class for Sri Lankan public holidays and Poya days
 * Includes Buddhist, Hindu, Islamic, and Christian holidays
 */
public class SriLankanHolidays {

    // Fixed holidays for 2025
    private static final Map<Integer, List<LocalDate>> FIXED_HOLIDAYS = new HashMap<>();
    
    // Poya days (Buddhist full moon holidays) for 2025-2026
    private static final Map<Integer, List<LocalDate>> POYA_DAYS = new HashMap<>();
    
    static {
        // Initialize 2025 fixed holidays
        List<LocalDate> holidays2025 = new ArrayList<>();
        holidays2025.add(LocalDate.of(2025, Month.JANUARY, 14));   // Tamil Thai Pongal Day
        holidays2025.add(LocalDate.of(2025, Month.FEBRUARY, 4));   // Independence Day
        holidays2025.add(LocalDate.of(2025, Month.MARCH, 14));     // Mahasivarathri Day
        holidays2025.add(LocalDate.of(2025, Month.APRIL, 10));     // Eid-Ul-Fitr
        holidays2025.add(LocalDate.of(2025, Month.APRIL, 11));     // Day prior to New Year
        holidays2025.add(LocalDate.of(2025, Month.APRIL, 12));     // Sinhala & Tamil New Year
        holidays2025.add(LocalDate.of(2025, Month.APRIL, 13));     // Day following New Year
        holidays2025.add(LocalDate.of(2025, Month.APRIL, 14));     // Good Friday
        holidays2025.add(LocalDate.of(2025, Month.MAY, 1));        // May Day
        holidays2025.add(LocalDate.of(2025, Month.MAY, 12));       // Vesak Poya
        holidays2025.add(LocalDate.of(2025, Month.MAY, 13));       // Day following Vesak
        holidays2025.add(LocalDate.of(2025, Month.JUNE, 16));      // Eid-Ul-Alha
        holidays2025.add(LocalDate.of(2025, Month.DECEMBER, 25));  // Christmas
        FIXED_HOLIDAYS.put(2025, holidays2025);
        
        // Initialize 2026 fixed holidays
        List<LocalDate> holidays2026 = new ArrayList<>();
        holidays2026.add(LocalDate.of(2026, Month.JANUARY, 14));   // Tamil Thai Pongal Day
        holidays2026.add(LocalDate.of(2026, Month.FEBRUARY, 4));   // Independence Day
        holidays2026.add(LocalDate.of(2026, Month.MARCH, 3));      // Mahasivarathri Day
        holidays2026.add(LocalDate.of(2026, Month.MARCH, 30));     // Eid-Ul-Fitr
        holidays2026.add(LocalDate.of(2026, Month.APRIL, 3));      // Good Friday
        holidays2026.add(LocalDate.of(2026, Month.APRIL, 13));     // Day prior to New Year
        holidays2026.add(LocalDate.of(2026, Month.APRIL, 14));     // Sinhala & Tamil New Year
        holidays2026.add(LocalDate.of(2026, Month.MAY, 1));        // May Day
        holidays2026.add(LocalDate.of(2026, Month.MAY, 31));       // Vesak Poya
        holidays2026.add(LocalDate.of(2026, Month.JUNE, 1));       // Day following Vesak
        holidays2026.add(LocalDate.of(2026, Month.JUNE, 6));       // Eid-Ul-Alha
        holidays2026.add(LocalDate.of(2026, Month.DECEMBER, 25));  // Christmas
        FIXED_HOLIDAYS.put(2026, holidays2026);
        
        // Initialize 2025 Poya days
        List<LocalDate> poya2025 = new ArrayList<>();
        poya2025.add(LocalDate.of(2025, Month.JANUARY, 13));   // Duruthu
        poya2025.add(LocalDate.of(2025, Month.FEBRUARY, 12));  // Navam
        poya2025.add(LocalDate.of(2025, Month.MARCH, 14));     // Medin
        poya2025.add(LocalDate.of(2025, Month.APRIL, 12));     // Bak
        poya2025.add(LocalDate.of(2025, Month.MAY, 12));       // Vesak
        poya2025.add(LocalDate.of(2025, Month.JUNE, 10));      // Poson
        poya2025.add(LocalDate.of(2025, Month.JULY, 10));      // Esala
        poya2025.add(LocalDate.of(2025, Month.AUGUST, 8));     // Nikini
        poya2025.add(LocalDate.of(2025, Month.SEPTEMBER, 7));  // Binara
        poya2025.add(LocalDate.of(2025, Month.OCTOBER, 6));    // Vap
        poya2025.add(LocalDate.of(2025, Month.NOVEMBER, 5));   // Il
        poya2025.add(LocalDate.of(2025, Month.DECEMBER, 4));   // Unduvap
        POYA_DAYS.put(2025, poya2025);
        
        // Initialize 2026 Poya days
        List<LocalDate> poya2026 = new ArrayList<>();
        poya2026.add(LocalDate.of(2026, Month.JANUARY, 2));    // Duruthu
        poya2026.add(LocalDate.of(2026, Month.FEBRUARY, 1));   // Navam
        poya2026.add(LocalDate.of(2026, Month.MARCH, 3));      // Medin
        poya2026.add(LocalDate.of(2026, Month.APRIL, 1));      // Bak
        poya2026.add(LocalDate.of(2026, Month.MAY, 1));        // Vesak
        poya2026.add(LocalDate.of(2026, Month.MAY, 30));       // Poson
        poya2026.add(LocalDate.of(2026, Month.JUNE, 29));      // Esala
        poya2026.add(LocalDate.of(2026, Month.JULY, 28));      // Nikini
        poya2026.add(LocalDate.of(2026, Month.AUGUST, 27));    // Binara
        poya2026.add(LocalDate.of(2026, Month.SEPTEMBER, 25)); // Vap
        poya2026.add(LocalDate.of(2026, Month.OCTOBER, 25));   // Il
        poya2026.add(LocalDate.of(2026, Month.NOVEMBER, 23));  // Unduvap
        poya2026.add(LocalDate.of(2026, Month.DECEMBER, 23));  // Duruthu
        POYA_DAYS.put(2026, poya2026);
    }
    
    /**
     * Get all holidays for a specific year
     */
    public static Set<LocalDate> getHolidaysForYear(int year) {
        Set<LocalDate> holidays = new HashSet<>();
        holidays.addAll(FIXED_HOLIDAYS.getOrDefault(year, Collections.emptyList()));
        holidays.addAll(POYA_DAYS.getOrDefault(year, Collections.emptyList()));
        return holidays;
    }
    
    /**
     * Check if a specific date is a holiday
     */
    public static boolean isHoliday(LocalDate date) {
        Set<LocalDate> holidays = getHolidaysForYear(date.getYear());
        return holidays.contains(date);
    }
    
    /**
     * Get all holidays within a date range
     */
    public static List<LocalDate> getHolidaysInRange(LocalDate start, LocalDate end) {
        List<LocalDate> holidays = new ArrayList<>();
        
        for (int year = start.getYear(); year <= end.getYear(); year++) {
            holidays.addAll(getHolidaysForYear(year));
        }
        
        return holidays.stream()
                .filter(date -> !date.isBefore(start) && !date.isAfter(end))
                .sorted()
                .collect(Collectors.toList());
    }
    
    /**
     * Calculate working days excluding holidays
     * Note: Saturdays and Sundays are working days in this organization
     * 
     * @param start Start date (inclusive)
     * @param end End date (inclusive)
     * @return Number of working days (total days - public holidays)
     */
    public static int calculateWorkingDays(LocalDate start, LocalDate end) {
        if (start.isAfter(end)) {
            throw new IllegalArgumentException("Start date must be before or equal to end date");
        }
        
        // Calculate total days (inclusive)
        long totalDays = java.time.temporal.ChronoUnit.DAYS.between(start, end) + 1;
        
        // Count holidays in the range
        long holidayCount = getHolidaysInRange(start, end).size();
        
        // Working days = total days - holidays (weekends are working days)
        return (int) (totalDays - holidayCount);
    }
    
    /**
     * Get a result object with detailed breakdown
     */
    public static WorkingDaysResult calculateWorkingDaysDetailed(LocalDate start, LocalDate end) {
        int totalDays = (int) java.time.temporal.ChronoUnit.DAYS.between(start, end) + 1;
        List<LocalDate> holidays = getHolidaysInRange(start, end);
        int workingDays = totalDays - holidays.size();
        
        return new WorkingDaysResult(totalDays, holidays.size(), workingDays, holidays);
    }
    
    /**
     * Result object containing working days calculation details
     */
    public static class WorkingDaysResult {
        private final int totalDays;
        private final int holidays;
        private final int workingDays;
        private final List<LocalDate> holidayList;
        
        public WorkingDaysResult(int totalDays, int holidays, int workingDays, List<LocalDate> holidayList) {
            this.totalDays = totalDays;
            this.holidays = holidays;
            this.workingDays = workingDays;
            this.holidayList = holidayList;
        }
        
        public int getTotalDays() { return totalDays; }
        public int getHolidays() { return holidays; }
        public int getWorkingDays() { return workingDays; }
        public List<LocalDate> getHolidayList() { return holidayList; }
    }
}

File: src/utils/sriLankanHolidays.js
/**
 * Sri Lankan Public Holidays Utility
 * 
 * Includes:
 * - Fixed holidays (e.g., Christmas, Independence Day)
 * - Poya days (calculated based on lunar calendar)
 * - Other government holidays
 */

// Fixed holidays for Sri Lanka (month is 0-indexed: 0=Jan, 11=Dec)
const FIXED_HOLIDAYS = {
  2025: [
    { date: '2025-01-14', name: 'Tamil Thai Pongal Day' },
    { date: '2025-02-04', name: 'Independence Day' },
    { date: '2025-03-14', name: 'Mahasivarathri Day' },
    { date: '2025-04-10', name: 'Eid-Ul-Fitr (Ramazan Festival Day)' },
    { date: '2025-04-11', name: 'Day prior to Sinhala & Tamil New Year Day' },
    { date: '2025-04-12', name: 'Sinhala & Tamil New Year Day' },
    { date: '2025-04-13', name: 'Day following Sinhala & Tamil New Year Day' },
    { date: '2025-04-14', name: 'Good Friday' },
    { date: '2025-05-01', name: 'May Day' },
    { date: '2025-05-12', name: 'Vesak Full Moon Poya Day' },
    { date: '2025-05-13', name: 'Day following Vesak Full Moon Poya Day' },
    { date: '2025-06-16', name: 'Eid-Ul-Alha (Hadji Festival Day)' },
    { date: '2025-12-25', name: 'Christmas Day' },
  ],
  2026: [
    { date: '2026-01-14', name: 'Tamil Thai Pongal Day' },
    { date: '2026-02-04', name: 'Independence Day' },
    { date: '2026-03-03', name: 'Mahasivarathri Day' },
    { date: '2026-03-30', name: 'Eid-Ul-Fitr (Ramazan Festival Day)' },
    { date: '2026-04-13', name: 'Day prior to Sinhala & Tamil New Year Day' },
    { date: '2026-04-14', name: 'Sinhala & Tamil New Year Day' },
    { date: '2026-04-03', name: 'Good Friday' },
    { date: '2026-05-01', name: 'May Day' },
    { date: '2026-05-31', name: 'Vesak Full Moon Poya Day' },
    { date: '2026-06-01', name: 'Day following Vesak Full Moon Poya Day' },
    { date: '2026-06-06', name: 'Eid-Ul-Alha (Hadji Festival Day)' },
    { date: '2026-12-25', name: 'Christmas Day' },
  ],
};

// Poya Days for 2025-2026 (Full moon days - Buddhist religious holidays)
const POYA_DAYS = {
  2025: [
    { date: '2025-01-13', name: 'Duruthu Full Moon Poya Day' },
    { date: '2025-02-12', name: 'Navam Full Moon Poya Day' },
    { date: '2025-03-14', name: 'Medin Full Moon Poya Day' },
    { date: '2025-04-12', name: 'Bak Full Moon Poya Day' },
    { date: '2025-05-12', name: 'Vesak Full Moon Poya Day' },
    { date: '2025-06-10', name: 'Poson Full Moon Poya Day' },
    { date: '2025-07-10', name: 'Esala Full Moon Poya Day' },
    { date: '2025-08-08', name: 'Nikini Full Moon Poya Day' },
    { date: '2025-09-07', name: 'Binara Full Moon Poya Day' },
    { date: '2025-10-06', name: 'Vap Full Moon Poya Day' },
    { date: '2025-11-05', name: 'Il Full Moon Poya Day' },
    { date: '2025-12-04', name: 'Unduvap Full Moon Poya Day' },
  ],
  2026: [
    { date: '2026-01-02', name: 'Duruthu Full Moon Poya Day' },
    { date: '2026-02-01', name: 'Navam Full Moon Poya Day' },
    { date: '2026-03-03', name: 'Medin Full Moon Poya Day' },
    { date: '2026-04-01', name: 'Bak Full Moon Poya Day' },
    { date: '2026-05-01', name: 'Vesak Full Moon Poya Day' },
    { date: '2026-05-30', name: 'Poson Full Moon Poya Day' },
    { date: '2026-06-29', name: 'Esala Full Moon Poya Day' },
    { date: '2026-07-28', name: 'Nikini Full Moon Poya Day' },
    { date: '2026-08-27', name: 'Binara Full Moon Poya Day' },
    { date: '2026-09-25', name: 'Vap Full Moon Poya Day' },
    { date: '2026-10-25', name: 'Il Full Moon Poya Day' },
    { date: '2026-11-23', name: 'Unduvap Full Moon Poya Day' },
    { date: '2026-12-23', name: 'Duruthu Full Moon Poya Day' },
  ],
};

/**
 * Get all holidays for a specific year
 * @param {number} year - The year to get holidays for
 * @returns {Array} Array of holiday objects with date and name
 */
export const getHolidaysForYear = (year) => {
  const fixed = FIXED_HOLIDAYS[year] || [];
  const poya = POYA_DAYS[year] || [];
  return [...fixed, ...poya];
};

/**
 * Get all holidays within a date range
 * @param {Date|string} startDate - Start date
 * @param {Date|string} endDate - End date
 * @returns {Array} Array of holiday objects within the range
 */
export const getHolidaysInRange = (startDate, endDate) => {
  const start = new Date(startDate);
  const end = new Date(endDate);
  
  const startYear = start.getFullYear();
  const endYear = end.getFullYear();
  
  let allHolidays = [];
  
  // Get holidays for all years in the range
  for (let year = startYear; year <= endYear; year++) {
    allHolidays = [...allHolidays, ...getHolidaysForYear(year)];
  }
  
  // Filter holidays within the date range
  return allHolidays.filter(holiday => {
    const holidayDate = new Date(holiday.date);
    return holidayDate >= start && holidayDate <= end;
  });
};

/**
 * Check if a specific date is a holiday
 * @param {Date|string} date - Date to check
 * @returns {Object|null} Holiday object if it's a holiday, null otherwise
 */
export const isHoliday = (date) => {
  const checkDate = new Date(date);
  const dateStr = checkDate.toISOString().split('T')[0];
  const year = checkDate.getFullYear();
  
  const allHolidays = getHolidaysForYear(year);
  return allHolidays.find(h => h.date === dateStr) || null;
};

/**
 * Calculate working days excluding holidays (but including weekends since employees work Sat/Sun)
 * @param {Date|string} startDate - Start date
 * @param {Date|string} endDate - End date
 * @returns {Object} Object with totalDays, holidays, workingDays, and holidaysList
 */
export const calculateWorkingDays = (startDate, endDate) => {
  const start = new Date(startDate);
  const end = new Date(endDate);
  
  // Calculate total days (inclusive)
  const totalDays = Math.floor((end - start) / (1000 * 60 * 60 * 24)) + 1;
  
  // Get holidays in the range
  const holidaysInRange = getHolidaysInRange(start, end);
  
  // Working days = total days - holidays (weekends are working days)
  const workingDays = totalDays - holidaysInRange.length;
  
  return {
    totalDays,
    holidays: holidaysInRange.length,
    workingDays: Math.max(0, workingDays),
    holidaysList: holidaysInRange,
  };
};

/**
 * Get holiday map for easy lookup (used by date picker)
 * @param {number} year - Year to get holidays for
 * @returns {Object} Map of date strings to holiday names
 */
export const getHolidayMap = (year) => {
  const holidays = getHolidaysForYear(year);
  const map = {};
  holidays.forEach(h => {
    map[h.date] = h.name;
  });
  return map;
};

/**
 * Get all holidays for multiple years (used for date range spanning years)
 * @param {number} startYear - Start year
 * @param {number} endYear - End year
 * @returns {Array} Array of all holidays
 */
export const getHolidaysForYearRange = (startYear, endYear) => {
  let allHolidays = [];
  for (let year = startYear; year <= endYear; year++) {
    allHolidays = [...allHolidays, ...getHolidaysForYear(year)];
  }
  return allHolidays;
};

const sriLankanHolidaysUtil = {
  getHolidaysForYear,
  getHolidaysInRange,
  isHoliday,
  calculateWorkingDays,
  getHolidayMap,
  getHolidaysForYearRange,
};

export default sriLankanHolidaysUtil;
